vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,876 @@
+    msg << "ns_counter_size " << ns_counter_size << "\n";
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&ns_counter_size), sizeof(ns_counter_size), read, msg, text);
+    for (auto& ns_count : ns_counter)
+    {
+      bin_text_read_write_fixed(model_file, (char*)&ns_count.first, sizeof(ns_count.first), read, msg, text);
+      bin_text_read_write_fixed(
+          model_file, reinterpret_cast<char*>(&ns_count.second), sizeof(ns_count.second), read, msg, text);
+    }
+
+    // Save index_queue
+    index_queue_size = index_queue.size();
+    msg << "index_queue_size " << index_queue_size << "\n";
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&index_queue_size), sizeof(index_queue_size), read, msg, text);
+    while (!index_queue.empty())
+    {
+      auto& pri_ind = index_queue.top();
+      bin_text_read_write_fixed(model_file, (char*)&pri_ind.first, sizeof(pri_ind.first), read, msg, text);
+      bin_text_read_write_fixed(model_file, (char*)&pri_ind.second, sizeof(pri_ind.second), read, msg, text);
+      index_queue.pop();
+    }
+  }
+}
+
+void interaction_config_manager::persist(metric_sink& metrics)
+{
+  metrics.int_metrics_list.emplace_back("test_county", total_learn_count);
+  metrics.int_metrics_list.emplace_back("current_champ", current_champ);
+  for (size_t live_slot = 0; live_slot < scores.size(); ++live_slot)
+  { scores[live_slot].persist(metrics, "_" + std::to_string(live_slot)); }
+}
+
+// This sets up example with correct ineractions vector
+void interaction_config_manager::apply_config(example* ec, size_t live_slot)
+{
+  if (ec == nullptr) { return; }
+  if (live_slot < max_live_configs) { ec->interactions = &(scores[live_slot].live_interactions); }
+  else
+  {
+    THROW("fatal (automl): trying to apply a config higher than max configs allowed");
+  }
+}
+
+void interaction_config_manager::revert_config(example* ec) { ec->interactions = nullptr; }
+
+template <typename CMType, bool is_explore>
+void predict_automl(automl<CMType>& data, multi_learner& base, multi_ex& ec)
+{
+  size_t champ_live_slot = data.cm->current_champ;
+  for (example* ex : ec) { data.cm->apply_config(ex, champ_live_slot); }
+
+  auto restore_guard = VW::scope_exit([&data, &ec, &champ_live_slot] {
+    for (example* ex : ec) { data.cm->revert_config(ex); }
+  });
+
+  base.predict(ec, champ_live_slot);
+}
+
+// inner loop of learn driven by # MAX_CONFIGS
+template <typename CMType>
+void automl<CMType>::offset_learn(multi_learner& base, multi_ex& ec, CB::cb_class& logged, size_t labelled_action)
+{
+  for (size_t live_slot = 0; live_slot < cm->scores.size(); ++live_slot)
+  {
+    for (example* ex : ec) { cm->apply_config(ex, live_slot); }
+
+    auto restore_guard = VW::scope_exit([this, &ec, &live_slot] {
+      for (example* ex : ec) { this->cm->revert_config(ex); }
+    });
+
+    if (!base.learn_returns_prediction) { base.predict(ec, live_slot); }
```

Comment:
Nice!

Reply:
this has always been like this, or what is the meaning of savings? Plus it doesn't matter since automl only has been tested with cb_explore_adf and I believe that one is set to false. So we still call predict, let me know if I'm out of date about this Griffin.

Reply:
No this looks good to me